vortex-file/src/generic.rs

```diff
@@ -89,100 +94,164 @@ pub struct GenericScanDriver<R> {
-            }
+        let inflight = inflight_segments.clone();
+        let segment_requests = segment_requests.filter_map(move |request| {
+            filter_with_cache(request, segment_cache.clone(), inflight.clone())
+        });
+        let segment_cache = self.segment_cache.clone();
+        let inflight = inflight_segments.clone();
+        let prefetch_stream = prefetch_stream.filter_map(move |request| {
+            filter_with_cache(request, segment_cache.clone(), inflight.clone())
         });
 
         // Grab all available segment requests from the I/O queue so we get maximal visibility into
         // the requests for coalescing.
         // Note that we can provide a somewhat arbitrarily high capacity here since we're going to
         // deduplicate and coalesce. Meaning the resulting stream will at-most cover the entire
         // file and therefore be reasonably bounded.
-        let io_stream = io_stream.ready_chunks(1024);
-
         // Coalesce the segment requests to minimize the number of I/O operations.
         let perf_hint = self.read.performance_hint();
-        let io_stream = io_stream
-            .map(move |r| coalesce(r, perf_hint.coalescing_window(), perf_hint.max_read()))
-            .flat_map(stream::iter)
-            .inspect(move |coalesced| self.metrics.record(coalesced));
+        let io_stream = SegmentRequestStream {
+            requests: segment_requests,
+            prefetch: prefetch_stream,
+            requests_ready_chunks: 1024,
+            prefetch_ready_chunks: self.options.io_concurrency,
+        }
+        .map(move |r| coalesce(r, perf_hint.coalescing_window(), perf_hint.max_read()))
+        .flat_map(stream::iter)
+        .inspect(move |(coalesced, _)| self.metrics.record(coalesced));
 
         // Submit the coalesced requests to the I/O.
         let read = self.read.clone();
         let segment_map = self.footer.segment_map().clone();
         let segment_cache = self.segment_cache.clone();
-        let io_stream = io_stream.map(move |request| {
+        let io_stream = io_stream.map(move |(request, cancellation_handle)| {
             let read = read.clone();
             let segment_map = segment_map.clone();
             let segment_cache = segment_cache.clone();
+            let inflight = inflight_segments.clone();
             async move {
-                evaluate(
-                    read.clone(),
-                    request,
-                    segment_map.clone(),
-                    segment_cache.clone(),
-                )
-                .await
+                select! {
+                    _ = cancellation_handle.cancelled().fuse() => Ok(()),
+                    evaluated = evaluate(
+                        read.clone(),
+                        request,
+                        segment_map.clone(),
+                        segment_cache.clone(),
+                        inflight.clone(),
+
+                    ).fuse() => evaluated,
+                }
```

Comment:
I wonder if we merge the cancellation_handle into request and propagate the cancellation from there though maybe there's too many checks for cancellation that way?

Reply:
CoalescedSegmentRequest is just a data object so this wouldn't change anything for the better